### PR TITLE
[mlir][SCF] Use Affine ops for indexing math.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
For index type of induction variable, the indexing math is better represented using affine ops such as `affine.delinearize_index`.

This also further demonstrates that some of these `affine` ops might need to move to a different dialect. For one these ops only support `IndexType` when they should be able to work with any integer type.

This change also includes some canonicalization patterns for `affine.delinearize_index` operation to
1) Drop unit `basis` values
2) Remove the `delinearize_index` op when the `linear_index` is a loop induction variable of a normalized loop and the `basis` is of size 1 and is also the upper bound of the normalized loop.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
THere are two commits in this PR. They could be reviewed independently.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 1]
```diff
@@ -714,9 +732,38 @@ Range mlir::emitNormalizedLoopBounds(RewriterBase &rewriter, Location loc,
...
+      preservedUses.insert(preservedUse);
+    }
+  }
+  rewriter.replaceAllUsesExcept(normalizedIv, denormalizedIvVal, preservedUses);
+}
+
 void mlir::denormalizeInductionVariable(RewriterBase &rewriter, Location loc,
                                         Value normalizedIv, OpFoldResult origLb,
                                         OpFoldResult origStep) {
+  if (getType(origLb) == rewriter.getIndexType()) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:763]: Can this be `getType(origLb).isIndex()`?

https://github.com/llvm/llvm-project/blob/de1f5b96adcea52bf7c9670c46123fe1197050d2/mlir/include/mlir/IR/Types.h#L127

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 2]
```diff
@@ -774,6 +840,16 @@ static Value getProductOfIntsOrIndexes(RewriterBase &rewriter, Location loc,
 static std::pair<SmallVector<Value>, SmallPtrSet<Operation *, 2>>
 delinearizeInductionVariable(RewriterBase &rewriter, Location loc,
                              Value linearizedIv, ArrayRef<Value> ubs) {
+
+  if (linearizedIv.getType() == rewriter.getIndexType()) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:844]: ditto: can we use `isIndex()`?

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp [hunk 3]
```diff
@@ -671,9 +672,26 @@ LogicalResult mlir::loopUnrollJamByFactor(scf::ForOp forOp,
...
+  AffineExpr e = (s1 - s0).ceilDiv(s2);
+  normalizedLoopBounds.size =
+      affine::makeComposedFoldedAffineApply(rewriter, loc, e, {lb, ub, step});
+  return normalizedLoopBounds;
+}
+
 Range mlir::emitNormalizedLoopBounds(RewriterBase &rewriter, Location loc,
                                      OpFoldResult lb, OpFoldResult ub,
                                      OpFoldResult step) {
+  if (getType(lb) == rewriter.getIndexType()) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:692]: ditto: can we use `isIndex()`?

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 4]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+        zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+      }
+      return zero.value();
+    };
+
+    // Replace all indices corresponding to unit-extent basis with 0.
+    // Remaining basis can be used to get a new `affine.delinearize_index` op.
+    SmallVector<Value> newOperands;
+    for (auto [index, basis] : llvm::enumerate(delinearizeOp.getBasis())) {
+      if (matchPattern(basis, m_One())) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4560]: use `isConstantIntValue(1)`?

https://github.com/llvm/llvm-project/blob/1e23a6142a827cda89fa4d8335afebd89701991d/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L93-L94C6

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 5]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+        replacements[index] = getZero();
+      } else {
+        newOperands.push_back(basis);
+      }
+    }
+
+    if (newOperands.size() == delinearizeOp.getBasis().size()) {
+      return failure();
+    }
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4569]: I know that you want to create the zero value only when it is easy. But would it be easier if you unconditionally create the zero value in the beginning and erase it here?

In this context, you don't need the lambda.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 6]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+
+    if (!newOperands.empty()) {
+      auto newDelinearizeOp = rewriter.create<affine::AffineDelinearizeIndexOp>(
+          loc, delinearizeOp.getLinearIndex(), newOperands);
+      int newIndex = 0;
+      // Map back the new delinearized indices to the values they replace.
+      for (auto i : llvm::seq<size_t>(0, replacements.size())) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4576]: can this just be `for (auto &replacement : replacements)`?

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 7]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+struct DropDelinearizeOfSingleLoop
+    : public OpRewritePattern<affine::AffineDelinearizeIndexOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(affine::AffineDelinearizeIndexOp delinearizeOp,
+                                PatternRewriter &rewriter) const override {
+    auto basis = delinearizeOp.getBasis();
+    if (basis.size() != 1) {
+      return failure();
+    }
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4613]: llvm style nit: remove braces for simple statement.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 8]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+    auto inductionVar = cast<BlockArgument>(delinearizeOp.getLinearIndex());
+    if (!inductionVar)
+      return failure();
+
+    // Check that the parent is a `LoopLikeOpInterface`.
+    auto loopLikeOp = cast<LoopLikeOpInterface>(
+        inductionVar.getParentRegion()->getParentOp());
+    if (!loopLikeOp) {
+      return failure();
+    }
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4625]: ditto: remove the braces.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 9]
```diff
@@ -4534,6 +4534,140 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
...
+    if (!upperBounds || upperBounds->size() != 1 ||
+        upperBounds->front() != getAsOpFoldResult(basis.front())) {
+      return rewriter.notifyMatchFailure(delinearizeOp,
+                                         "`basis` is not upper bound");
+    }
+
+    // Check that the lower bound is zero.
+    auto lowerBounds = loopLikeOp.getLoopLowerBounds();
+    if (!lowerBounds || lowerBounds->size() != 1 ||
+        !isConstantIntValue(lowerBounds->front(), 0)) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4647]: optional nit: this can be `!isZeroIndex(lowerBounds->front())` which is shorter.

**Review by hanhanW — COMMENTED** [review]
Overall looks good, just some nits

**Review by hanhanW — APPROVED** [review]

**Pull request merged**